pkg/controllers/scheduling/scheduling_controller.go

```diff
@@ -0,0 +1,226 @@
+	clusterlisterv1alpha1 "github.com/open-cluster-management/api/client/cluster/listers/cluster/v1alpha1"
+	clusterapiv1 "github.com/open-cluster-management/api/cluster/v1"
+	clusterapiv1alpha1 "github.com/open-cluster-management/api/cluster/v1alpha1"
+)
+
+const (
+	clusterSetLabel = "cluster.open-cluster-management.io/clusterset"
+	placementLabel  = "cluster.open-cluster-management.io/placement"
+)
+
+var ResyncInterval = 2 * time.Minute
+
+// schedulingController schedules cluster decisions for Placements
+type schedulingController struct {
+	clusterClient           clusterclient.Interface
+	clusterLister           clusterlisterv1.ManagedClusterLister
+	clusterSetLister        clusterlisterv1alpha1.ManagedClusterSetLister
+	clusterSetBindingLister clusterlisterv1alpha1.ManagedClusterSetBindingLister
+	placementLister         clusterlisterv1alpha1.PlacementLister
+	placementDecisionLister clusterlisterv1alpha1.PlacementDecisionLister
+	scheduleFunc            scheduleFunc
+}
+
+// NewDecisionSchedulingController return an instance of schedulingController
+func NewSchedulingController(
+	clusterClient clusterclient.Interface,
+	clusterLister clusterlisterv1.ManagedClusterLister,
+	clusterSetLister clusterlisterv1alpha1.ManagedClusterSetLister,
+	clusterSetBindingLister clusterlisterv1alpha1.ManagedClusterSetBindingLister,
+	placementInformer clusterinformerv1alpha1.PlacementInformer,
+	placementDecisionInformer clusterinformerv1alpha1.PlacementDecisionInformer,
+	recorder events.Recorder,
+) factory.Controller {
+	// build controller
+	c := schedulingController{
+		clusterClient:           clusterClient,
+		clusterLister:           clusterLister,
+		clusterSetLister:        clusterSetLister,
+		clusterSetBindingLister: clusterSetBindingLister,
+		placementLister:         placementInformer.Lister(),
+		placementDecisionLister: placementDecisionInformer.Lister(),
+		scheduleFunc:            schedule,
+	}
+
+	return factory.New().
+		WithInformersQueueKeyFunc(func(obj runtime.Object) string {
+			key, _ := cache.MetaNamespaceKeyFunc(obj)
+			return key
+		}, placementInformer.Informer()).
+		WithFilteredEventsInformersQueueKeyFunc(func(obj runtime.Object) string {
+			accessor, _ := meta.Accessor(obj)
+			labels := accessor.GetLabels()
+			placementName := labels[placementLabel]
+			return fmt.Sprintf("%s/%s", accessor.GetNamespace(), placementName)
+		}, func(obj interface{}) bool {
+			accessor, err := meta.Accessor(obj)
+			if err != nil {
+				return false
+			}
+			labels := accessor.GetLabels()
+			if _, ok := labels[placementLabel]; ok {
+				return true
+			}
+			return false
+		}, placementDecisionInformer.Informer()).
+		// TODO: monitor more resources, like clusters, clustersets and clustersetbindings
+		WithSync(c.sync).
+		ResyncEvery(ResyncInterval).
+		ToController("SchedulingController", recorder)
+}
+
+func (c *schedulingController) sync(ctx context.Context, syncCtx factory.SyncContext) error {
```

Comment:
followup: You need to ensure that an error to sync a particular placement results in a condition of some kind being placed on the placement.  The structure of returning an error here is good.  I suspect you need to expand the factory to add a sync error handler.